cicerone/spec/models.py

```diff
@@ -0,0 +1,351 @@
+            if method in data:
+                operations[method] = Operation.from_dict(method.upper(), path, data[method])
+
+        return cls(path=path, operations=operations)
+
+
+class Paths(BaseModel):
+    """Container for all path items in the spec."""
+
+    model_config = {"extra": "allow"}
+
+    items: dict[str, PathItem] = Field(default_factory=dict)
+
+    def __getitem__(self, path: str) -> PathItem:
+        """Get a path item by path string."""
+        return self.items[path]
+
+    def __contains__(self, path: str) -> bool:
+        """Check if a path exists."""
+        return path in self.items
+
+    def all_operations(self) -> Generator[Operation, None, None]:
+        """Yield all operations across all paths."""
+        for path_item in self.items.values():
+            yield from path_item.operations.values()
+
+    @classmethod
+    def from_dict(cls, data: Mapping[str, Any]) -> "Paths":
+        """Create Paths from a dictionary."""
+        items = {}
+        for path, path_data in data.items():
+            if isinstance(path_data, dict):
+                items[path] = PathItem.from_dict(path, path_data)
+        return cls(items=items)
+
+
+class Components(BaseModel):
+    """Container for reusable component definitions."""
+
+    model_config = {"extra": "allow"}
+
+    schemas: dict[str, Schema] = Field(default_factory=dict)
+
+    def get(self, schema_name: str) -> Schema | None:
+        """Get a schema by name."""
+        return self.schemas.get(schema_name)
+
+    @classmethod
+    def from_spec(cls, raw: Mapping[str, Any], version: Version) -> "Components":
+        """Create Components from spec data, handling both OpenAPI 3.x and Swagger 2.0."""
+        schemas = {}
+
+        # OpenAPI 3.x: components.schemas
+        if version.major >= 3 and "components" in raw and "schemas" in raw["components"]:
+            for name, schema_data in raw["components"]["schemas"].items():
+                schemas[name] = Schema.from_dict(schema_data)
+
+        # Swagger 2.0: definitions
+        elif version.major == 2 and "definitions" in raw:
+            for name, schema_data in raw["definitions"].items():
+                schemas[name] = Schema.from_dict(schema_data)
+
+        return cls(schemas=schemas)
+
+
+class OpenAPISpec(BaseModel):
+    """Top-level OpenAPI specification model."""
+
+    model_config = {"extra": "allow", "arbitrary_types_allowed": True}
+
+    raw: dict[str, Any]
+    version: Version
+    paths: Paths
+    components: Components
+
+    @model_validator(mode="before")
+    @classmethod
+    def convert_version(cls, data: Any) -> Any:
+        """Convert version string to Version object if needed."""
+        if isinstance(data, dict) and "version" in data and isinstance(data["version"], str):
+            data["version"] = Version(data["version"])
+        return data
+
+    @classmethod
+    def from_dict(cls, data: Mapping[str, Any]) -> "OpenAPISpec":
```

Comment:
Let's move all the `from_` parsing methods into their own module under `cicerone/parse` or something. I'd like the objects to be as pure as possible and there is some cross over here I'm not happy with.